File: numericslib.py
# -*- coding: utf-8 -*-
"""
Created on Sun Apr  9 12:37:12 2023

@author: 12019044

Solver Functions
"""
import numpy as np


def gradient_flux(scalar_function, x, y, z):
    return 0.5 * np.array([scalar_function[z, y, x + 1] - scalar_function[z, y, x - 1], 
                           scalar_function[z, y + 1, x] - scalar_function[z, y - 1, x], 
                           scalar_function[z + 1, y, x] - scalar_function[z - 1, y, x]])
    
def divergence_flux(vector_function, x, y, z):
    return 0.5 * (vector_function[z, y, x + 1, 0] - vector_function[z, y, x - 1, 0] + 
                  vector_function[z, y + 1, x, 1] - vector_function[z, y - 1, x, 1] +
                  vector_function[z + 1, y, x, 2] - vector_function[z - 1, y, x, 2])
    
def curl_flux(vector_function, x, y, z):
    return 0.5 * np.array([vector_function[z, y + 1, x, 2] - vector_function[z, y - 1, x, 2]
                         - vector_function[z + 1, y, x, 1] + vector_function[z - 1, y, x, 1],
                           vector_function[z + 1, y, x, 0] - vector_function[z - 1, y, x, 0]
                         - vector_function[z, y, x + 1, 2] + vector_function[z, y, x - 1, 2],
                           vector_function[z, y, x + 1, 1] - vector_function[z, y, x - 1, 1]
                         - vector_function[z, y + 1, x, 0] + vector_function[z, y - 1, x, 0]])

def dot_product(vector_a, vector_b):
    return (vector_a[1] * vector_b[1] + vector_a[2] * vector_b[2] + vector_a[3] * vector_b[3])

def cross_product(vector_a, vector_b):
    return np.array(vector_a[2] * vector_b[3] - vector_a[3] * vector_b[2],
                    vector_a[3] * vector_b[1] - vector_a[1] * vector_b[3],
                    vector_a[1] * vector_b[2] - vector_a[2] * vector_b[1])


def gradient_flux_2d(scalar_function, x, y):
    return 0.5 * np.array([scalar_function[y, x + 1] - scalar_function[y, x - 1], 
                           scalar_function[y + 1, x] - scalar_function[y - 1, x]])
    
def divergence_flux_2d(vector_function, x, y):
    return 0.5 * (vector_function[y, x + 1, 0] - vector_function[y, x - 1, 0] + 
                  vector_function[y + 1, x, 1] - vector_function[y - 1, x, 1])
    
def curl_flux_2d(vector_function, x, y):
    return 0.5 * (vector_function[y, x + 1, 1] - vector_function[y, x - 1, 1]
                - vector_function[y + 1, x, 0] + vector_function[y - 1, x, 0])

File: Numerics.NET/NumericsEngine/NumericsLib.cs
using System;

namespace NumericsEngine
{
    public static class UniversalConstants
    {
        public const float pi = 3.141592653589793f;

        public const float e_0 = 8.8541878128f * 0.000000000001f; //10^-12
        public const float e_0Inverse = 1.1294090674f * 100000000000; // 10^11

        public const float m_0 = 1.2566370621f * 0.000001f; //10^-6
        public const float m_0Inverse = 7.9577471502f * 100000; //10^5

        public const float c = 299792458;
        public const float cSquared = c * c;

        public const float electronCharge = 1.602176634f * 0.0000000000000000001f; //10^-19
    }

    public static class NumericsLib
    {
        public static double[] GradientFlux(double[,,] ScalarFunction, int x, int y, int z)
        {
            return new double[3] { 0.5*(ScalarFunction[z, y, x + 1] - ScalarFunction[z, y, x - 1]),
                                   0.5*(ScalarFunction[z, y + 1, x] - ScalarFunction[z, y - 1, x]),
                                   0.5*(ScalarFunction[z + 1, y, x] - ScalarFunction[z - 1, y, x])};
        }

        public static double DivergenceFlux(double[,,,] VectorFunction, int x, int y, int z)
        {
            return 0.5 * (VectorFunction[z, y, x + 1, 0] - VectorFunction[z, y, x - 1, 0] +
                          VectorFunction[z, y + 1, x, 1] - VectorFunction[z, y - 1, x, 1] +
                          VectorFunction[z + 1, y, x, 2] - VectorFunction[z - 1, y, x, 2]);
        }

        public static double[] CurlFlux(double[,,,] VectorFunction, int x, int y, int z)
        {
            return new double[3] { 0.5*(VectorFunction[z, y + 1, x, 2] - VectorFunction[z, y - 1, x, 2]
                                      - VectorFunction[z + 1, y, x, 1] + VectorFunction[z - 1, y, x, 1]),
                                   0.5*(VectorFunction[z + 1, y, x, 0] - VectorFunction[z - 1, y, x, 0]
                                      - VectorFunction[z, y, x + 1, 2] + VectorFunction[z, y, x - 1, 2]),
                                   0.5*(VectorFunction[z, y, x + 1, 1] - VectorFunction[z, y, x - 1, 1]
                                      - VectorFunction[z, y + 1, x, 0] + VectorFunction[z, y - 1, x, 0])};
        }

        public static double[] GradientFlux2D(double[,] ScalarFunction, int x, int y)
        {
            return new double[2] { 0.5*(ScalarFunction[y, x + 1] - ScalarFunction[y, x - 1]),
                                   0.5*(ScalarFunction[y + 1, x] - ScalarFunction[y - 1, x])};
        }

        public static double DivergenceFlux2D(double[,,] VectorFunction, int x, int y)
        {
            return 0.5 * (VectorFunction[y, x + 1, 0] - VectorFunction[y, x - 1, 0] +
                          VectorFunction[y + 1, x, 1] - VectorFunction[y - 1, x, 1]);
        }

        public static double CurlFlux2D(double[,,] VectorFunction, int x, int y)
        {
            return  0.5*(VectorFunction[y, x + 1, 1] - VectorFunction[y, x - 1, 1]
                       - VectorFunction[y + 1, x, 0] + VectorFunction[y - 1, x, 0]);
        }


    }
}
